subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java

```diff
@@ -16,20 +16,41 @@
 
 package org.gradle.internal;
 
+import com.google.common.collect.Lists;
 import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.lang.StringUtils;
 import org.gradle.api.GradleException;
 import org.gradle.api.UncheckedIOException;
 
 import java.io.File;
 import java.io.IOException;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 import java.util.Iterator;
-import java.util.LinkedList;
 import java.util.List;
 
 public class FileUtils {
     public static final int WINDOWS_PATH_LIMIT = 260;
 
+    private static final Comparator<File> FILE_SEGMENT_COMPARATOR = new Comparator<File>() {
+        @Override
+        public int compare(File left, File right) {
+            String[] leftSegments = StringUtils.split(left.getPath(), File.separator);
```

Comment:
I feel that using StringUtils.split() makes the code pretty hard to reason when the file path is not an absolute path, and the code may not work as expected in certain cases.

Considering  a method call like
```java
FileUtils.calculateRoots(Arrays.asList(new File("/a"), new File("a"), new File("/a/a")))
```
An reasonable output should be Arrays.asList(new File("/a"), new File("a")), but this implementation generates Arrays.asList(new File("/a"), new File("a"), new File("/a/a")) instead. The reason is, this comparator considers new File("/a") and new File("a") as equal so that these two elements will stay in their original order after sorting. As the result, new File("/a") and new File("/a/a") won't be "merged" correctly.

I think the approach pointed out by @tjni would solve this problem and is also pretty easy to understand. Converting all files to absolute files may also solve this issue, but I am not sure if there are other edge cases left.